python/aws_lambda_powertools/tracing/base.py

```diff
@@ -0,0 +1,233 @@
+
+    @abc.abstractmethod
+    def patch(self, modules: List[str] = None):
+        """Patch modules for instrumentation
+
+        If modules are None, it should patch
+        all supported modules by the provider.
+
+        Parameters
+        ----------
+        modules : List[str], optional
+            List of modules to be pathced, by default None
+            e.g. `['boto3', 'requests']`
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def create_subsegment(self, name: str):
+        """Creates subsegment/span with a given name
+
+        Parameters
+        ----------
+        name : str
+            Subsegment/span name
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def end_subsegment(self):
+        """Ends an existing subsegment"""
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def put_metadata(self, key: str, value: Any, namespace: str = None):
+        """Adds metadata to existing segment/span or subsegment
+
+        Parameters
+        ----------
+        key : str
+            Metadata key
+        value : Any
+            Metadata value
+        namespace : str, optional
+            Metadata namespace, by default None
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def put_annotation(self, key: str, value: Any):
+        """Adds annotation/label to existing segment/span or subsegment
+
+        Parameters
+        ----------
+        key : str
+            Annotation/label key
+        value : Any
+            Annotation/label value
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def disable_tracing_provider(self):
+        """Forcefully disables tracing provider"""
+        raise NotImplementedError
+
+
+class XrayProvider(TracerProvider):
+    """X-Ray Tracer provider
+
+    It implements all basic ``aws_lambda_powertools.tracing.base.TracerProvider` methods,
+    and automatically annotates cold start on first subsegment created.
+
+    Parameters
+    ----------
+    client : aws_xray_sdk.core.xray_recorder
+        X-Ray recorder client
+    """
+
+    def __init__(self, client: aws_xray_sdk.core.xray_recorder = aws_xray_sdk.core.xray_recorder):
+        self.client = client
+
+    def create_subsegment(self, name: str) -> aws_xray_sdk.core.models.subsegment:
+        """Creates subsegment/span with a given name
+
+        Parameters
+        ----------
+        name : str
+            Subsegment name
+
+        Example
+        -------
+
+        **Creates a subsegment**
+
+            self.create_subsegment(name="a meaningful name")
+
+        Returns
+        -------
+        aws_xray_sdk.core.models.subsegment
+            AWS X-Ray Subsegment
+        """
+        # Will no longer be needed once #155 is resolved
+        # https://github.com/aws/aws-xray-sdk-python/issues/155
+        subsegment = self.client.begin_subsegment(name=name)
+        global is_cold_start
```

Comment:
It makes sense to have a global variable in the context of Lambda execution environments, but wouldn't it make more sense to have it at the class level? (think `self.__class__.__is_cold_start`)

Reply:
If they instantiate Tracer inside the handler then it'll always be cold start, hence why we have in the module